src/devices/bus/epson_qx/multifont.cpp

```diff
@@ -0,0 +1,258 @@
+	device_option_expansion_interface(mconfig, *this),
+	m_mcu(*this, "mcu"),
+	m_i8243(*this, "i8243"),
+	m_fonts(*this, "font%u",0),
+	m_iobase(*this, "IOBASE")
+{
+}
+
+//-------------------------------------------------
+//  device_input_ports - device-specific ports
+//-------------------------------------------------
+ioport_constructor multifont_device::device_input_ports() const
+{
+	return INPUT_PORTS_NAME(multifont);
+}
+
+//-------------------------------------------------
+//  device_rom_region - device-specific roms
+//-------------------------------------------------
+const tiny_rom_entry *multifont_device::device_rom_region() const
+{
+	return ROM_NAME(multifont);
+}
+
+//-------------------------------------------------
+//  device_add_mconfig - device-specific config
+//-------------------------------------------------
+void multifont_device::device_add_mconfig(machine_config &config)
+{
+	I8039(config, m_mcu, 9_MHz_XTAL);
+	m_mcu->set_addrmap(AS_PROGRAM, &multifont_device::rom_map);
+	m_mcu->prog_out_cb().set(m_i8243, FUNC(i8243_device::prog_w));
+	m_mcu->p1_out_cb().set(FUNC(multifont_device::p1_w));
+	m_mcu->p2_out_cb().set(FUNC(multifont_device::p2_w));
+	m_mcu->bus_out_cb().set(FUNC(multifont_device::write_bus));
+	m_mcu->bus_in_cb().set(FUNC(multifont_device::read_bus));
+	m_mcu->t1_in_cb().set(FUNC(multifont_device::t1_r));
+	m_mcu->t0_in_cb().set(FUNC(multifont_device::t0_r));
+	I8243(config, m_i8243);
+	m_i8243->p4_out_cb().set(FUNC(multifont_device::rom_addr_w<0>));
+	m_i8243->p5_out_cb().set(FUNC(multifont_device::rom_addr_w<4>));
+	m_i8243->p6_out_cb().set(FUNC(multifont_device::rom_addr_w<8>));
+	m_i8243->p7_out_cb().set(FUNC(multifont_device::rom_select_w));
+}
+
+//-------------------------------------------------
+//  device_start - device-specific startup
+//-------------------------------------------------
+void multifont_device::device_start()
+{
+}
+
+//-------------------------------------------------
+//  device_reset - device-specific reset
+//-------------------------------------------------
+void multifont_device::device_reset()
+{
+	m_data_out = 0x00;
+	m_data_in = 0x00;
+	m_p2 = 0x00;
+	m_status = 0x00;
+	address_space &space = m_bus->iospace();
+
+	offs_t	iobase = (m_iobase->read() & 0x06) + 0xf8;
+
+	space.install_device(iobase, iobase+1, *this, &multifont_device::map);
+}
+
+uint8_t multifont_device::t0_r()
+{
+	return m_status & 0x01;
+}
+
+uint8_t multifont_device::t1_r()
+{
+	return (m_status & 0x80) >> 7;
+}
+
+void multifont_device::write_bus(uint8_t data)
+{
+	m_status = (m_status & 0x7f) | 0x80;
+	m_data_out = data;
+}
+
+uint8_t multifont_device::read_bus()
+{
+	if (m_p1 & 0x01) {
+		m_status = (m_status & 0xfe);
+		return m_data_in;
+	} else if (m_p1 & 0x02) {
+		uint8_t *rom = m_fonts[m_rom_bank]->base();
+		return rom[m_address];
+	}
+	return 0xff;
+}
+
+void multifont_device::p1_w(uint8_t data)
+{
+	m_p1 = data;
+	if ((m_p1 & 0x80) == 0) {
+		get_slot()->intl_w(ASSERT_LINE);
+	}
```

Comment:
This isn’t a strobed port.  Should this be detecting a falling edge to assert the interrupt?

Reply:
I don't think so. The schematic shows that pin 7 of port 1 is connected to the preset pin of a flip flop the output of which controls if the interrupt is asserted or not. So the way I read it the interrupt should be sent immediately without needing any edge detection.

Reply:
If it’s connected directly to the preset pin of a flip-flop, it will prevent the interrupt from being acknowledged.  Are you taking that into consideration?

Reply:
Yeah the flop flop is cleared by doing a read from the status register, doing so on the schematic creates a clock pulse on the clock pin and the data pin is connected to ground. In the code i just clear the interrupt line whenever the status register is read.

Reply:
I just had a look at the schematic, you’ll need to check to ensure P7 isn’t preventing the interrupt from being acknowledged.

Reply:
Can you explain how P1.7 would be preventing interrupt acknowledgement? I've been looking at it bu fail to see what the issue here is.

Reply:
If P1.7 is being held low, the flipflop will ignore its clock input, hence preventing an interrupt from being acknowledged.  To make the emulation accurate, you need to check that P1.7 isn’t being held low when the host reads from the card as that would prevent the interrupt output from being deasserted.